### PR TITLE
Theme Showcase: Do not show wporg themes if they are also retired wpcom themes

### PR DESCRIPTION
**Pull request by taipeicoder**
Related to #78091

## Proposed Changes

The current implementation shows a wporg theme even if its wpcom counterpart is retired. For instance, the theme Maxwell exists in both wporg and wpcom, but it's retired in wpcom. This leads to a scenario where the Theme Showcase will show the wporg version of the theme, but when the user clicks on the theme, the detail page will say that the theme is retired. This behavior is confusing, and might lead to issues as described in #78091.

This PR updates the logic so that retired wpcom themes are always omitted from the Theme Showcase, even if their wporg counterparts exist. In other words, when users search for Maxwell, even though the theme exists in wporg, it will not be shown in the Theme Showcase listing because it's retired in wpcom.

## Testing Instructions

* Head to the Theme Showcase.
* Search for a retired theme that also exists in wporg, such as Maxwell or Adelle.
* Ensure that the theme is not shown in the theme listing.

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by arthur791004** [comment]
Tested but the Maxwell theme is `null` so the activate action is still there

image

File: client/state/themes/selectors/is-theme-retired.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+/**
+ * Whether a WPCOM theme given by its ID is retired.
+ *
+ * @param  {Object} state   Global state tree
+ * @param  {string} themeId Theme ID
+ * @returns {boolean}        True if the theme is retired
+ */
+export function isThemeRetired( state, themeId ) {
+	return isRetired( getTheme( state, 'wpcom', themeId ) );
```
Comment (arthur791004) [inline client/state/themes/selectors/is-theme-retired.js]: The theme might not be fetched. So, one option is to get theme from wporg as well and add the slug to `RETIRED_THEME_SLUGS_SET`

Comment (taipeicoder) [reply client/state/themes/selectors/is-theme-retired.js]: Hmmm the `retired` property is only from wpcom, which is why we only query for the theme in wpcom. How can I reproduce the theme being `null` 🤔 ?

Comment (arthur791004) [reply client/state/themes/selectors/is-theme-retired.js]: The theme was null when I went to the Theme Showcase and searched “Maxwell”. Or maybe you can head to the search result page directly 🤔

Comment (taipeicoder) [reply client/state/themes/selectors/is-theme-retired.js]: @arthur791004 with D113775-code merged, the query will no longer return `null`. Could you test again? Thanks!

Comment (arthur791004) [reply client/state/themes/selectors/is-theme-retired.js]: We have to pass `retired=true` to themes endpoint to get the result

Comment (taipeicoder) [reply client/state/themes/selectors/is-theme-retired.js]: Oops, I forgot to commit 😅

**Review by arthur791004 — APPROVED** [review]
Tested again and looks good to me 👍

**Comment by arthur791004** [comment]
BTW, D113775-code has a side effect on the search result...we have to hide the retired themes from wpcom on client side as well

image

**Comment by taipeicoder** [comment]
> BTW, D113775-code has a side effect on the search result...we have to hide the retired themes from wpcom on client side as well
> 
> image

Was your search term "blank canvas". I could only see the one in wpcom 🤔

Screenshot 2023-06-16 at 7 00 44 PM

Were you using the calypso.live link?

**Comment by arthur791004** [comment]
> Were you using the calypso.live link?

Yes, I was using the calypso.live link and searched for the “blank canvas”. 

**Here is the result on production**

image

**Here is the result from the endpoint**

image

**Comment by taipeicoder** [comment]
@arthur791004 I've changed the PR proposal to an approach that I think makes more sense, and updated the description accordingly. I've moved the PR to Needs Review again 🙂

File: client/my-sites/themes/helpers.js [hunk 2]
```diff
@@ -162,7 +163,12 @@ export function interlaceThemes( wpComThemes, wpOrgThemes, searchTerm, isLastPag
 	const restWpComThemes = matchingTheme
 		? wpComThemes.filter( ( theme ) => theme.id !== matchingTheme.id )
 		: wpComThemes;
-	interlacedThemes.push( ...restWpComThemes );
+
+	// The themes endpoint returns retired themes when search term exists.
+	// See: https://github.com/Automattic/wp-calypso/pull/78231
+	interlacedThemes.push(
+		...( searchTerm ? restWpComThemes.filter( ( theme ) => ! theme.retired ) : restWpComThemes )
```
Comment (arthur791004) [inline client/my-sites/themes/helpers.js:170]: [nit] We can filter out the retired themes regardless of the search term. But the current way is also fine 👍

**Review by arthur791004 — APPROVED** [review]
Tested again and looks good 🎉

image

image

File: client/my-sites/themes/helpers.js (hunk 2 above)
Comment (taipeicoder) [reply client/my-sites/themes/helpers.js:170]: I'll leave the logic in for now, since I like the idea of the small performance gain! 🙂

**Pull request merged**